pgjdbc/src/main/java/org/postgresql/util/PGServiceFile.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2018, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.util;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.nio.charset.StandardCharsets;
+import java.text.MessageFormat;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Scanner;
+
+/*
+ * See https://www.postgresql.org/docs/current/static/libpq-pgservice.html.
+ */
+public class PGServiceFile {
+  public static Map load(String service) throws Exception {
+    String filename = findPath();
+    PGServiceFile file = new PGServiceFile();
+    Scanner in = new Scanner(new FileInputStream(filename), StandardCharsets.UTF_8.name());
+    try {
+      file.parse(in);
+    } finally {
+      in.close();
+    }
+    return file.getService(service);
+  }
+
+  private static String findPath() {
+    String filename = System.getProperty("org.postgresql.pgservicefile");
+    if (filename == null) {
+      filename = System.getenv().get("PGSERVICEFILE");
+    }
+    if (filename == null) {
+      filename = System.getProperty("user.home") + File.separator + ".pgservices.conf";
+    }
+    return filename;
+  }
+
+  private Map<String, Map<String, String>> sections;
```

Comment:
final?

Reply:
Sorry, i'm not a java developer, Can you elaborate on this?

Reply:
final just means that it won't be modified again. The map can be modified but you can not replace the map with a new map. Seems reasonable ```private final Map ...```